src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/DefaultMonitorService.java

```diff
@@ -70,37 +72,50 @@ public DefaultMonitorService(Log log) {
 
   @Override
   public MonitorConnectionContext startMonitoring(
-      String node,
+      Set<String> nodes,
       HostInfo hostInfo,
       PropertySet propertySet,
       int failureDetectionTimeMillis,
       int failureDetectionIntervalMillis,
       int failureDetectionCount) {
 
-    final IMonitor monitor = MONITOR_MAP.computeIfAbsent(
-        node,
-        k -> monitorInitializer.createMonitor(hostInfo, propertySet));
+    IMonitor monitor = null;
```

Comment:
As discussed offline, here are the propose changes to simply the code a little bit.

```Java
  @Override
  public MonitorConnectionContext startMonitoring(
      Set<String> nodeKey,
      HostInfo hostInfo,
      PropertySet propertySet,
      int failureDetectionTimeMillis,
      int failureDetectionIntervalMillis,
      int failureDetectionCount) {

    if (nodeKey.isEmpty()) {
      // exception because nodeKey should at least have `this.hostInfo.getHost()`
    }

    IMonitor monitor = null;
    for (String s : nodeKey) {
      monitor = MONITOR_MAP.get(s);
    }

    if (monitor == null) {
      monitor = MONITOR_MAP.computeIfAbsent(nodeKey.iterator().next(), k -> monitorInitializer.createMonitor(hostInfo, propertySet));
    }

    for (String s : nodeKey) {
      MONITOR_MAP.putIfAbsent(s, monitor);
    }

    if (threadPool == null) {
      threadPool = executorServiceInitializer.createExecutorService();
    }

    final MonitorConnectionContext context = new MonitorConnectionContext(
        nodeKey,
        log,
        failureDetectionTimeMillis,
        failureDetectionIntervalMillis,
        failureDetectionCount);

    monitor.startMonitoring(context);
    TASKS_MAP.putIfAbsent(monitor, threadPool.submit(monitor));

    return context;
  }
```

Reply:
After some thought, we can improve the above code using Java Stream.
```java
  @Override
  public MonitorConnectionContext startMonitoring(
      Set<String> nodeKey,
      HostInfo hostInfo,
      PropertySet propertySet,
      int failureDetectionTimeMillis,
      int failureDetectionIntervalMillis,
      int failureDetectionCount) {

    if (nodeKey.isEmpty()) {
      // exception because nodeKey should at least have `this.hostInfo.getHost()`
    }

    final IMonitor monitor;
    final Optional<String> node = nodeKey.stream().filter(MONITOR_MAP::containsKey).findFirst();
    if (node.isPresent()) {
      monitor = MONITOR_MAP.get(node.get());
    } else {
      monitor = MONITOR_MAP.computeIfAbsent(nodeKey.iterator().next(), k -> monitorInitializer.createMonitor(hostInfo, propertySet));
    }

    for (String s : nodeKey) {
      MONITOR_MAP.putIfAbsent(s, monitor);
    }

    if (threadPool == null) {
      threadPool = executorServiceInitializer.createExecutorService();
    }

    final MonitorConnectionContext context = new MonitorConnectionContext(
        nodeKey,
        log,
        failureDetectionTimeMillis,
        failureDetectionIntervalMillis,
        failureDetectionCount);

    monitor.startMonitoring(context);
    TASKS_MAP.putIfAbsent(monitor, threadPool.submit(monitor));

    return context;
  }
```

If you want to check whether the stream works, here is a simple code you can run to test it out.
```java
  public static void main(String[] args) {
    final Map<String, String> map = new ConcurrentHashMap<>();
    final Set<String> stringSet = new HashSet<>();

    map.put("1", "val1");
    map.put("2", "val2");
    map.put("3", "val3");

//    stringSet.add("1");
//    stringSet.add("3");
    stringSet.add("4");
    stringSet.add("5");

    final Optional<String> test = stringSet.stream().filter(map::containsKey).findFirst();
    if (test.isPresent()) {
      System.out.println(test.get());
      System.out.println(map.get(test.get()));
    } else {
      System.out.println(map.computeIfAbsent(stringSet.iterator().next(), k -> "hello"));
    }
  }
```

Reply:
I like a version with stream. It's cleaner and more readable. ^

Reply:
I'd also suggest to extract a logic that look for monitor object and update monitor map into a separate method. This logic is quite critical and we need unit tests for it.